subprojects/core/src/integTest/groovy/org/gradle/api/tasks/CopyTaskIntegrationSpec.groovy

```diff
@@ -2239,4 +2242,121 @@ class CopyTaskIntegrationSpec extends AbstractIntegrationSpec {
         "rename(Transformer)"       | "rename(org.gradle.internal.Transformers.noOpTransformer())"
     }
 
+    // region duplicates in compressed files
+    def "encountering duplicates in a zipTree vs an unzipped dir with DuplicateStrategy.FAIL should give a clear error"() {
+        given: "a directory with a file"
+        TestFile unzippedDir = file("before/files")
+        unzippedDir.file("sub/c.txt").createFile()
+
+        and: "a zip file containing it"
+        TestFile zipFile = file("before/files.zip")
+        unzippedDir.zipTo(zipFile)
+
+        and: "a copy task that copies from both of these, failing on duplicates"
+        buildFile << """
+            task (copy, type: Copy) {
+                from '${separatorsToUnix(unzippedDir.path)}'
+                from zipTree('${separatorsToUnix(zipFile.path)}')
+                into 'after'
+                duplicatesStrategy = DuplicatesStrategy.FAIL
+            }
+        """
+
+        expect: "a failure"
+        fails 'copy'
+
+        and: "with a clear failure message"
+        failure.assertHasCause('Encountered duplicate path "sub/c.txt", extracted from "files.zip", during copy operation configured with DuplicatesStrategy.FAIL')
```

Comment:
🤔 Usually we use single quotes to highlight paths/names.  

🤔 Does the path to the zip file need to be a fully qualified path? You could have a zip coming from Gradle user home or another task's build output. This makes it look like it might be in the root.

Reply:
Rewriting message.